### PR TITLE
improve polars query

### PR DESCRIPTION
**Pull request by ritchie46**

### CODE REVIEW

File: python_benchmarks/20_polars_lazy.py [hunk 1]
```diff
@@ -1,18 +1,15 @@
...
 def polars_lazy(fname):
-    dtypes = {f"column{i}": (polars.UInt32 if i == 0 else polars.Float32())
+    dtypes = {f"column{i}": (pl.UInt32 if i == 0 else pl.Float32())
               for i in range(9)}
-    return (polars.scan_csv(fname, dtypes=dtypes)
-                  .select(
-                      polars.sum_horizontal(polars.all()
-                                                  .exclude("0")
+    return (pl.scan_csv(fname, dtypes=dtypes)
+                  .filter(pl.sum_horizontal(pl.all()
```
Comment (ritchie46) [inline python_benchmarks/20_polars_lazy.py:8]: Doing this in the `filter` allows Polars to push this predicate completely down to the scan level. And is more similar to what is done in the SQL query. E.g. the full predicate is in the `where` clause. So it is fair to give polars that information as well.

File: python_benchmarks/20_polars_lazy.py [hunk 2]
```diff
@@ -1,18 +1,15 @@
...
+                                                  .exclude("column0")
                                                   .pow(2)
-                      ).sqrt()
-                       .alias("sum_sq"))
-                  .filter(polars.col("sum_sq") < 1e16)
-                  .count()
+                      ).sqrt() < 1e16)
+                  .select(pl.len())
```
Comment (ritchie46) [inline python_benchmarks/20_polars_lazy.py:12]: This is Polars' `SELECT count(*) ...` query, and that way we are not forced to materialize the other columns. (currently we still do, but are not forced).

File: python_benchmarks/21_polars_streaming.py [hunk 3]
```diff
@@ -1,20 +1,17 @@
...
-                                                  .pow(2)
-                      ).sqrt()
-                       .alias("sum_sq"))
-                  .filter(polars.col("sum_sq") < 1e16)
-                  .count()
-                  .collect(streaming=True)
-    )
+    return (pl.scan_csv(fname, dtypes=dtypes)
+            .filter(pl.sum_horizontal(pl.all()
+                                      .exclude("column0")
```
Comment (ritchie46) [inline python_benchmarks/21_polars_streaming.py:9]: Column `0` doesn't exist so Polars did more work than other solutions here as it also included the id column

**Review by ritchie46 — COMMENTED** [review]
Made changes to make it more similar to what is done in the SQL queries.

Half of the execution time is in the compute of the predicate. I am not sure we are super optimal here, as csv-parsing should be much more expensive.

**Comment by datapythonista** [comment]
This is great feedback, thank you! Not sure why I did in two steps instead of directly writing the `filter` as you did. The `exclude` not raising with a wrong column name seems like a bug to me. Maybe I'm missing something, but I can open an issue if you want.

Not sure if you run the benchmarks, but I tested your implementation, and both implementations run faster, very similar to the DuckDB time, and still one order of magnitude more than DataFusion:

```
pure_python_int,3.3809654712677
pure_python_float,3.78913950920105
pandas_c,1.5084028244018555
pandas_python,8.409186601638794
pandas_pyarrow,0.3095574378967285
pandas_pyarrow_arrow,0.28849101066589355
polars_lazy,0.1080479621887207
polars_streaming,0.11388874053955078
duckdb_sql,0.10797977447509766
datafusion_sql,0.0014221668243408203
numpy_loadtxt,1.834737777709961
```

I'll continue working on this, my main goal is not to just to compare the times, but to understand and write a post on why an implementation gets faster, particularly on the parallelization, which seems to be the trickiest part. I'll let you know of any finding that can be useful to you.

**Comment by ritchie46** [comment]
That's strange. On my machine datafusion is at least two times slower:

```
(venv) [filter_polars]  ritchie46 /home/ritchie46/Downloads/bench_csv$ time python python_benchmarks/20_polars_lazy.py                                                                               (base) 
shape: (1, 1)
┌─────┐
│ len │
│ --- │
│ u32 │
╞═════╡
│ 65  │
└─────┘

________________________________________________________
Executed in  114.44 millis    fish           external
   usr time  556.50 millis    0.00 micros  556.50 millis
   sys time   50.76 millis  871.00 micros   49.89 millis

(venv) [filter_polars]  ritchie46 /home/ritchie46/Downloads/bench_csv$ time python python_benchmarks/40_datafusion_sql.py                                                                            (base) 
DataFrame()
+----------+
| COUNT(*) |
+----------+
| 65       |
+----------+

________________________________________________________
Executed in  251.82 millis    fish           external
   usr time    1.31 secs    409.00 micros    1.31 secs
   sys time    1.36 secs    113.00 micros    1.36 secs


```

What is your hardware?

**Comment by datapythonista** [comment]
We are measuring different things, datafusion is slower when executing the whole script (including the imports), but it's extremely fast when timing only the call to the function.

**Comment by ritchie46** [comment]
If I do that Polars is 3 times faster

```
(venv) [filter_polars]  ritchie46 /home/ritchie46/Downloads/bench_csv[1] $ python python_benchmarks/20_polars_lazy.py                                                                                (base) 
shape: (1, 1)
┌─────┐
│ len │
│ --- │
│ u32 │
╞═════╡
│ 65  │
└─────┘
0.03916668891906738
(venv) [filter_polars]  ritchie46 /home/ritchie46/Downloads/bench_csv$ python python_benchmarks/40_datafusion_sql.py                                                                                 (base) 
DataFrame()
+----------+
| COUNT(*) |
+----------+
| 65       |
+----------+
0.12256932258605957

```


I wrapped the call like this: 

```python
(venv) [filter_polars]  ritchie46 /home/ritchie46/Downloads/bench_csv[1] $ python python_benchmarks/20_polars_lazy.py                                                                                (base) 
shape: (1, 1)
┌─────┐
│ len │
│ --- │
│ u32 │
╞═════╡
│ 65  │
└─────┘
0.03916668891906738
(venv) [filter_polars]  ritchie46 /home/ritchie46/Downloads/bench_csv$ python python_benchmarks/40_datafusion_sql.py                                                                                 (base) 
DataFrame()
+----------+
| COUNT(*) |
+----------+
| 65       |
+----------+
0.12256932258605957
```

Still curious about your hardware. :P

**Comment by ritchie46** [comment]
I have found it. Datafusion doesn't execute the query until you print the result. Adding a `print` to `__main__.py` shows that DataFusion is suddenly two orders of magnitude slower than before:

```
(venv) [filter_polars]  ritchie46 /home/ritchie46/Downloads/bench_csv$ python -m python_benchmarks                                                                                                   (base) 
shape: (1, 1)
┌─────┐
│ len │
│ --- │
│ u32 │
╞═════╡
│ 65  │
└─────┘
polars_lazy,0.04633188247680664
shape: (1, 1)
┌─────┐
│ len │
│ --- │
│ u32 │
╞═════╡
│ 65  │
└─────┘
polars_streaming,0.04948592185974121
DataFrame()
+----------+
| COUNT(*) |
+----------+
| 65       |
+----------+
datafusion_sql,0.13802528381347656

```

**Comment by datapythonista** [comment]
Ok, very silly mistake from my side. The datafusion benchmark doesn't materialize when running it with all other benchmarks. Only when calling the sprint, or in all the tests I did where I was printing the result to make sure it was doing what it was expected. Sorry for the noise, and thanks a lot for the help. I'll fix it now and update.

My hardware is a laptop Dell XPS running Arch Linux:
- Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
- RAM: MemTotal: 15815304 kB, Type: LPDDR3, Type Detail: Synchronous Unbuffered (Unregistered), Speed: 2133 MT/s
- Disk: KXG50ZNV512G NVMe TOSHIBA 512GB with ext4 partition

**Comment by ritchie46** [comment]
Yes, we came to the same conclusion. :smile:

**Comment by datapythonista** [comment]
Oh, your last message just loaded now...

You'll be happy to know that the same happens with DuckDB and it's way slower than Polars and DataFusion. Still having a look at it, as I don't see a `.collect()` in their API, and I'm just casting to string which may not be the fastest.

**Comment by ritchie46** [comment]
Yes, I am not sure, but I believe that duckdb will LIMIT the query to the amount it has to visualize. Which makes these benchmarks a bit harder to isolate.

**Comment by datapythonista** [comment]
I updated this repo, I also added the Polars SQL API, which happens to be the fastest.

**Comment by ritchie46** [comment]
Great!. Yes, if you do the SQL API you spell out the arithmetic. This allows Polars to do arithmetic on the same buffers. I have added that optimization upstream for `sum_horizontal` as well: 

https://github.com/pola-rs/polars/pull/14525

**Pull request closed without merging**